### PR TITLE
feat: classify DAG

### PR DESCRIPTION
**Pull request by kobayu858**
## Description

- Classification of DAG for deadline calculations.

## Related links

- original paper [OPEN_LINK](https://drive.google.com/file/d/1YH-MAWsw1VAvxqco4Beg3njQ9ykG-sb-/view?usp=drive_link) p5
- paper summary [EmbIV_INTERNAL_LINK](https://docs.google.com/presentation/d/1rabhhA9ziIcBjvT8XPLue78XEWcKMvH4/edit?usp=drive_link&ouid=114908281052440997053&rtpof=true&sd=true) p27 - p34
- design specifications [EmbIV INTERNAL LINK](https://docs.google.com/presentation/d/14YO4wHY1QqGOjoY9_g4djvtTaBXV0RCj/edit?usp=drive_link&ouid=114908281052440997053&rtpof=true&sd=true) p2, p4

## Pre-review checklist for the PR author

- [x] I've confirmed the [self-checklist](https://docs.google.com/spreadsheets/d/1p5KMcm752iCySHJz0xL1qf6fcLDKprbo/edit#gid=863618628).
- [x] The PR is ready for review.

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR has been properly tested.
- [x] The PR has been reviewed.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [x] There are no open discussions.
- [x] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

### CODE REVIEW

File: 2014_TPDS_basic_decomposition_based_algorithm/src/handle_segment.rs [hunk 1]
```diff
@@ -50,13 +57,55 @@ pub fn create_segments(dag: &mut Graph<NodeData, i32>) -> Vec<Segment> {
...
+                segment.parallel_degree += 1;
             }
         }
     }
 
     segments
 }
 
+fn classify_segment(volume: f32, period: f32, crit_path_len: f32, segment: &mut Segment) {
+    if segment.nodes.is_empty() {
```
Comment (atsushi421) [inline 2014_TPDS_basic_decomposition_based_algorithm/src/handle_segment.rs:69]: use assert.

File: 2014_TPDS_basic_decomposition_based_algorithm/src/handle_segment.rs [hunk 2]
```diff
@@ -50,13 +57,55 @@ pub fn create_segments(dag: &mut Graph<NodeData, i32>) -> Vec<Segment> {
...
+    segment.classification =
+        if segment.parallel_degree as f32 > volume / ((2.0 * period) - crit_path_len) {
+            Some(SegmentClassification::Heavy)
+        } else {
+            Some(SegmentClassification::Light)
+        };
+}
+
+#[allow(dead_code)] //TODO: remove
+fn classify_segments(
```
Comment (atsushi421) [inline 2014_TPDS_basic_decomposition_based_algorithm/src/handle_segment.rs:82]:
```suggestion
fn classify_dag(
```

File: 2014_TPDS_basic_decomposition_based_algorithm/src/handle_segment.rs [hunk 3]
```diff
@@ -50,13 +57,55 @@ pub fn create_segments(dag: &mut Graph<NodeData, i32>) -> Vec<Segment> {
...
+    for segment in segments.iter_mut() {
+        classify_segment(volume, period, crit_path_len, segment);
+    }
+
+    let (mut heavy_count, mut light_count) = (0, 0);
+    for segment in segments {
+        match segment.classification {
+            Some(SegmentClassification::Heavy) => heavy_count += 1,
+            Some(SegmentClassification::Light) => light_count += 1,
+            None => unreachable!("Segment classification error"),
```
Comment (atsushi421) [inline 2014_TPDS_basic_decomposition_based_algorithm/src/handle_segment.rs:97]:
```suggestion
            _ => unreachable!("Segment classification error"),
```

**Review by atsushi421 — CHANGES_REQUESTED** [review]

File: 2014_TPDS_basic_decomposition_based_algorithm/src/handle_segment.rs (hunk 2 above)
Comment (kobayu858) [reply 2014_TPDS_basic_decomposition_based_algorithm/src/handle_segment.rs:82]: I forgot to change it. Thank you very much.

File: 2014_TPDS_basic_decomposition_based_algorithm/src/handle_segment.rs (hunk 3 above)
Comment (kobayu858) [reply 2014_TPDS_basic_decomposition_based_algorithm/src/handle_segment.rs:97]: Is the reason for `_` for scalability?
Changed.

File: 2014_TPDS_basic_decomposition_based_algorithm/src/handle_segment.rs (hunk 1 above)
Comment (kobayu858) [reply 2014_TPDS_basic_decomposition_based_algorithm/src/handle_segment.rs:69]: Changed.

**Review by atsushi421 — APPROVED** [review]
LGTM

**Pull request merged**